### PR TITLE
[CWS] Add ebpfless functional tests

### PR DESCRIPTION
**Pull request by spikat**
### What does this PR do?

It bootstraps the functional tests for ebpfless mode.

Tl;dr:
* Added ebpfless tests:
- * 30 tests are passing.
- * 36 are skipped because we lack some support.
- * 1 is failing (but should not).
* Test fixed: 6
* Ebpfless fixes: 4
* Gaps identified: 25

To do so:
* I've added an async connection mode to the ptracer (aka `cws-instrumentation trace` cmd). Now with a `--async`, it will directly start to trace the program, and wait for a client to start sending pending messages.
* module_tester was also adapted: it could globally be configured in ebpfless mode if env `EBPFLESS` env variable is set; allowing to run the already existing epbf tests in ebpfless mode (but a new config option is also avaible to build specific ebpfless tests if needed).
* Test module reloading (between tests) has been disable, only the first instance will be up. This is an ~hard limitation: if we reload the module, we'll reset the ptracer connection and lose the process cache. Depending of future tests, this could become an issue.

Functional tests fixed:
* By checking the timeout error:
- * TestOpenDiscarded
- * TestProcessContext/args-overflow-list-50
- * TestProcessContext/args-overflow-list-500
- * TestProcessContext/envs-overflow-list-50
- * TestProcessContext/args-overflow-list-500
* TestProcessPIDVariable: remove wrong error check

Ebpfless other fixes/improvements:
* Exit events were caught but not sent. Now they are. This is a first basic implementation and lacks of context (cause, exit code and timestamp)
* Fix `execveat` segfault when passing an empty path (special case that was not handled)
* Fix chained exec events no different threads (where only the last exec was sent multiple times)
* Fix create_at field wich was missing (to handle secl rules based on it)

Remaining gaps to have a full open/exec coverage:
* Add proper timestamps to exec/exit events
* Add more context to exits (cause and return code)
* Checks on lineage were skipped (the fact to always have parent/ancestors and finish on PID 1). We may want to let them skipped, or revisit more globally ebpfless process contexts.
* Same for "hostname", as we can't determine it.
* Event (exec/exit/open) schema validation were also skipped (because a lot of fields are actually missing). We could either wait to add them all, or work with another temporary ebpfless schema. TODO: ask backend if receiving ebpfless events with unvalid schema trigger errors.
* Inodes checks are skipped. We should consider providing them and add the checks back
* Support SYS_CREAT
* Support SYS_TRUNCATE
* Support open_by_handle_at
* Support io_uring (maybe impossible?)
* MTime/CTime
* uid/gid for open files events
* Add user/group
* Follow links?
* args/envs truncation
* tty
* scoped variable and event timestamp (which should had work ootb)
* fsuid/fsgid
* be able to wrote rules on setuid/setgid (should work except for fs[ug]id).
* capset rules
* is_thread
* busybox special cases
* fileless / memfd
* other easy FIM syscalls: mkdir, rename, unlink, close, ...
* kill action

TODO: add a gitlab job

### Motivation

* Validate what we already built
* Have automatic tests for future improvements

### Additional Notes

Also, while trying to make most of the open/exec tests pass in ebpfless, I encountered some tests that were broken.
I guess that if we take a closer look on all other existing tests we shall have more surprises.

### Describe how to test/QA your changes

First, build the cws-instrumentation tool:

> inv -e cws-instrumentation.build

Then launch the ebpf testsuite:

> inv -e security-agent.ebpfless-functional-tests --skip-linters --testflags "-test.v"

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/security/ptracer/cws.go [hunk 1]
```diff
@@ -492,25 +553,44 @@ func StartCWSPtracer(args []string, probeAddr string, creds Creds, verbose bool)
...
-				// write size
-				var size [4]byte
-				native.Endian.PutUint32(size[:], uint32(len(data)))
-				if _, err = client.Write(size[:]); err != nil {
-					logErrorf("unabled to send size: %v", err)
-				}
-
-				if _, err = client.Write(data); err != nil {
-					logErrorf("unabled to send message: %v", err)
+				if clientReady {
```
Comment (safchain) [inline pkg/security/ptracer/cws.go:557]: what about just using the msgChan with a bigger size ? Here I have the impression that you can have out of order messages between open and exec events. We could put the clientReady in the "initGRPCcon" function (need to add a mutex)

File: pkg/security/resolvers/process/resolver_ebpfless.go [hunk 2]
```diff
@@ -101,7 +101,7 @@ func (p *EBPFLessResolver) AddExecEntry(key CacheResolverKey, file string, argv
 
 	entry.Process.ArgsEntry = &model.ArgsEntry{Values: argv}
 	if len(argv) > 0 {
-		entry.Process.Comm = argv[0]
+		entry.Process.Comm = filepath.Base(file)
```
Comment (safchain) [inline pkg/security/resolvers/process/resolver_ebpfless.go:104]: I'm not sure about that change. We need to check why you had to change it

File: pkg/security/tests/process_test.go [hunk 3]
```diff
@@ -361,6 +381,13 @@ func TestProcessContext(t *testing.T) {
 	})
 
 	test.Run(t, "args-overflow-single", func(t *testing.T, kind wrapperType, cmdFunc func(cmd string, args []string, envs []string) *exec.Cmd) {
+		if kind == dockerWrapperType && test.opts.staticOpts.enableEBPFLess == true {
```
Comment (safchain) [inline pkg/security/tests/process_test.go:384]: Per the "if" below it this that we skip whatever the "kind". I would suggest to remove this one

File: pkg/security/tests/process_test.go [hunk 4]
```diff
@@ -405,6 +432,12 @@ func TestProcessContext(t *testing.T) {
 	})
 
 	test.Run(t, "args-overflow-list-50", func(t *testing.T, kind wrapperType, cmdFunc func(cmd string, args []string, envs []string) *exec.Cmd) {
+		if kind == dockerWrapperType && test.opts.staticOpts.enableEBPFLess == true {
+			t.Skip("docker tests not supported")
+		}
+		if test.opts.staticOpts.enableEBPFLess == true {
```
Comment (safchain) [inline pkg/security/tests/process_test.go:438]: ditto

File: pkg/security/tests/process_test.go [hunk 5]
```diff
@@ -443,9 +476,19 @@ func TestProcessContext(t *testing.T) {
...
+		if err != nil {
+			t.Error(err)
+		}
 	})
 
 	test.Run(t, "args-overflow-list-500", func(t *testing.T, kind wrapperType, cmdFunc func(cmd string, args []string, envs []string) *exec.Cmd) {
+		if kind == dockerWrapperType && test.opts.staticOpts.enableEBPFLess == true {
+			t.Skip("docker tests not supported")
+		}
+		if test.opts.staticOpts.enableEBPFLess == true {
```
Comment (safchain) [inline pkg/security/tests/process_test.go:488]: ditto

File: pkg/security/tests/process_test.go [hunk 6]
```diff
@@ -139,11 +144,11 @@ func TestProcessContext(t *testing.T) {
 		},
 		{
 			ID:         "test_rule_ancestors_glob",
-			Expression: `exec.file.name == "ls" && exec.argv == "glob" && process.ancestors.file.path =~ "/usr/**"`,
+			Expression: `exec.file.name == "ls" && exec.argv == "glob" && process.ancestors.file.path in [~"/home/**", ~"/usr/**"]`,
```
Comment (safchain) [inline pkg/security/tests/process_test.go:147]: wondering why you need to add "home" here ?

File: pkg/security/tests/process_test.go [hunk 7]
```diff
@@ -139,11 +144,11 @@ func TestProcessContext(t *testing.T) {
...
 		},
 		{
 			ID:         "test_self_exec",
-			Expression: `exec.file.name == "syscall_tester" && exec.argv0 == "selfexec123" && process.comm == "exe"`,
+			Expression: `exec.file.name in ["syscall_tester", "exe"] && exec.argv0 == "selfexec123" && process.comm == "exe"`,
```
Comment (safchain) [inline pkg/security/tests/process_test.go:151]: Wondering why you need "exe" here ?

File: pkg/security/resolvers/process/resolver_ebpfless.go (hunk 2 above)
Comment (spikat) [reply pkg/security/resolvers/process/resolver_ebpfless.go]: It was a needed change to make the test `TestProcessContext/self-exec-std` pass. It uses selftest to self exec (using /proc/self/exe). It could be updated later if we add a proc fs fallback to catch the "real" comm. Since then, I think this change is ok

File: pkg/security/tests/process_test.go (hunk 6 above)
Comment (spikat) [reply pkg/security/tests/process_test.go:147]: Yep .. It's a bit hacky, I will try to fix more properly this test.
The issue is that assumes that, `testsuite` have at least one ancestors present in /usr/** (like bash?). But with ebpfless, testsuite comes as the top lineage process (so, as it stands in ~/go/src/github[...]/tests/, I've added `/home/**` but definitively it's not rock solid.

File: pkg/security/tests/process_test.go (hunk 7 above)
Comment (spikat) [reply pkg/security/tests/process_test.go:151]: In ebpfless, the reported file name of an execve of `/proc/self/exe` will be `exe`. To make this test pass as is, it would require to follow symlinks (which is not the case now). Would you prefer to skip this test instead (until we add symlink follow) ?

File: pkg/security/tests/process_test.go (hunk 3 above)
Comment (spikat) [reply pkg/security/tests/process_test.go:384]: The first `if` is to skip ebpfless docker tests, the second is temporary until we add what's missing ebpfless side to make it pass (and still, only on host).

**Review by spikat — COMMENTED** [review]
Most of the comments has been addressed on the last commit. Some are still open to discuss. Please have another look @safchain :)

File: pkg/security/tests/process_test.go (hunk 6 above)
Comment (spikat) [reply pkg/security/tests/process_test.go]: I've reworked the test on my last PR, using syscall tester as a proxy to execute ls, and we are now searching into `/tmp/***` instead (used to write the syscall tester on disk before executing it).

File: pkg/security/tests/module_tester.go [hunk 8]
```diff
@@ -1089,6 +1131,17 @@ func newTestModule(t testing.TB, macroDefs []*rules.MacroDefinition, ruleDefs []
 		t.Logf("%s entry stats: %s\n", t.Name(), GetEBPFStatusMetrics(testMod.probe))
 	}
 
+	if opts.staticOpts.enableEBPFLess {
+		t.Logf("EBPFLess mode, waiting for a client to connect\n")
+		for i := 0; i < 120; i++ {
```
Comment (safchain) [inline pkg/security/tests/module_tester.go:1136]: should we use a retry.Do here and return an error if still 0 connection after 120 seconds ?

Comment (spikat) [reply pkg/security/tests/module_tester.go]: Indeed! Updated on last commit, thanks

**Review by safchain — APPROVED** [review]

**Review by safchain — APPROVED** [review]

**Review by safchain — APPROVED** [review]
